### PR TITLE
Replace all item flag constants with static string_ids (FINAL part 5 of #44694)

### PR DESCRIPTION
**Pull request by Aivean**
#### Summary

SUMMARY: Infrastructure "Speed up item processing by changing item flag constants to static string_ids"

#### Purpose of change

This is the final part of #44694. 

#### Describe the solution

In this PR:

* replaced string arguments of `has_flag` and similar methods with (effectively) static `flag_str_id` constants
* changed type arguments of several methods, like `Character::exclusive_flag_coverage` or `inventory::leak_level`
* removed adapter methods that took strings instead of `flag_id`

Although the amount of lines changed is large, 99% of the changes are along the lines of:
```cpp
it->set_flag( "FILTHY" );
```
into
```cpp
it->set_flag( flag_FILTHY );
```

#### Describe alternatives you've considered

Alternatives are outlined  here: #44694. 

#### Testing

* Ensured that game compiles and loads without warnings
* Smoke checked several scenarios that involve flags:
    * saving and loading items with flags
    * modifying clothes
    * debug-changing item flags
    * allergies still work

### CODE REVIEW

**Review by mlangsdorf — COMMENTED** [review]
I reviewed basecamp, suffer, and the vehicle changes and noted one change.

I am not happy with stuffing all the static definitions in flag.h, because that creates another wildly included header that will force large scale recompilation when it changes.

File: src/vehicle.cpp [hunk 1]
```diff
@@ -4937,7 +4937,7 @@ void vehicle::power_parts()
             const int gen_energy_bat = power_to_energy_bat( part_epower_w( elem ), 1_turns );
             if( parts[ elem ].is_unavailable() ) {
                 continue;
-            } else if( parts[ elem ].info().has_flag( flag_PERPETUAL ) ) {
+            } else if( parts[ elem ].info().has_flag( STATIC( std::string( "PERPETUAL" ) ) ) ) {
```
[the remark this answers is not in the document]
Comment (Aivean) [reply src/vehicle.cpp:4940]: Yes, there's an item flag "PERPETUAL" (used for fuel and ammo) which has now `flag_str_id<json_flag>` type. There is also a vehicle part flag "PERPETUAL". I haven't touched vehicle part flags, they are still strings.

**Comment by Aivean** [comment]
> I reviewed basecamp, suffer, and the vehicle changes and noted one change.
> 
> I am not happy with stuffing all the static definitions in flag.h, because that creates another wildly included header that will force large scale recompilation when it changes.

Thank you for the review.

Yes, I also thought about that and I agree that in the future flags will have to be decentralized. However, specifically for this PR that presents a different problem:
1. Lots of string flags previously had following name format:  `static std::string flag_FLAG_NAME("FLAG_NAME")`
2. We have a clang-tidy check in place that enforces following format for static string_ids: 
      `static flag_str_id json_flag_FLAG_NAME("FLAG_NAME")`
3. I was asked by @KorGgenT previously to reduce the scope of the changes as much as possible, and [was specifically asked to try and not rename static constants](https://github.com/CleverRaven/Cataclysm-DDA/pull/44694#issuecomment-705125066).
4. Renaming existing usage of `flag_XXX` constants **will double the amount of lines changed in this PR**.

Note, I already tried to avoid including "flag.h" when the number of flags used was small (i.e. ≈1-5).

TL;DR:  I can see only one way to address your concern: move flag constants from flag.h into other files. That will require marking them as `static` and, because of the clang-tidy check, renaming them to have `json_flag_` prefix.  Consequently that will double the amount of lines changed in this PR and delay the review.

What we also can do, is merge this as is (and temporarily suffer the excessive header inclusion) and then I can replace that inclusion with local static definitions (adding `json_flag_` prefix) on file-by-file basis.

@mlangsdorf , what do you think?  Are there other options that I missed?

**Comment by jbytheway** [comment]
FYI, it's easy to edit the clang-tidy check to change the enforced prefix string for a particular `string_id` type if that would be helpful.
